services/github/github-directory-file-count.service.js

```diff
@@ -0,0 +1,105 @@
+'use strict'
+
+const path = require('path')
+const Joi = require('joi')
+const { InvalidParameter } = require('..')
+const { ConditionalGithubAuthV3Service } = require('./github-auth-service')
+const { documentation, errorMessagesFor } = require('./github-helpers')
+
+const schema = Joi.alternatives(
+  Joi.object({}).required(),
+  Joi.array()
+    .items(
+      Joi.object({
+        path: Joi.string().required(),
+        type: Joi.string().required(),
+      })
+    )
+    .required()
+)
+
+const queryParamSchema = Joi.object({
+  type: Joi.any().valid('dir', 'file'),
+  extension: Joi.string(),
+})
+
+module.exports = class GithubDirectoryFileCount extends ConditionalGithubAuthV3Service {
+  static category = 'size'
+
+  static route = {
+    base: 'github/directory-file-count',
+    pattern: ':user/:repo/:path*',
+    queryParamSchema,
+  }
+
+  static examples = [
+    {
+      title: 'GitHub repo directory file count',
+      pattern: ':user/:repo/:path',
+      namedParams: { user: 'badges', repo: 'shields', path: 'services' },
+      staticPreview: this.render({ count: 10 }),
+      documentation,
+    },
+    {
+      title: 'GitHub repo directory file count (dir type)',
+      pattern: ':user/:repo/:path',
+      namedParams: { user: 'badges', repo: 'shields', path: 'services' },
+      queryParams: { type: 'dir' },
+      staticPreview: this.render({ count: 8 }),
+      documentation,
+    },
+    {
+      title: 'GitHub repo directory file count (file type)',
+      pattern: ':user/:repo/:path',
+      namedParams: { user: 'badges', repo: 'shields', path: 'services' },
+      queryParams: { type: 'file' },
+      staticPreview: this.render({ count: 2 }),
+      documentation,
+    },
+    {
+      title: 'GitHub repo directory file count (file extension)',
+      pattern: ':user/:repo/:path',
+      namedParams: { user: 'badges', repo: 'shields', path: 'services' },
+      queryParams: { extension: 'js' },
+      staticPreview: this.render({ count: 1 }),
+      documentation,
+    },
+  ]
+
+  static defaultBadgeData = { color: 'blue', label: 'files' }
+
+  static render({ count }) {
+    return {
+      message: count,
+    }
+  }
+
+  async fetch({ user, repo, path = '' }) {
+    return this._requestJson({
+      url: `/repos/${user}/${repo}/contents/${path}`,
+      schema,
+      errorMessages: errorMessagesFor('repo or directory not found'),
+    })
+  }
+
+  static transform(files, { type, extension }) {
+    if (type) {
+      files = files.filter(file => file.type === type)
+    }
```

Comment:
Technically a user could provide a `type` of `dir` but also provide an `extension` param value. In this case I feel like we should either throw an `InvalidParameter` error outright, or return early since there's not really a need to iterate through the collection again to further filter by file extension when the collection only consists of directories

Reply:
I intentionally did not restrict the use of extension to `file` type only. I thought someone might find it helpful to be able to filter directories by extension if needed. I have never come across such a need but I thought leaving it will keep code simpler and might help someone. Since I don't have strong feelings about this I can change the code as you suggested.

Reply:
Within the context of a GitHub repo, what will qualify as a folder/directory extension? Wouldn't this just functionally be a filter of directories that happened to have a name ending with some `.foobar`?

I do have some worries that this endpoint in particular could be abused, so if we can prevent an extra loop over the response, and processing step for each entry, then I'd really like to do so, especially since there's not a strong valid use case for it.

Reply:
Exactly, it would boil down to filtering directories with extension. I have restricted the use of `extension` parameter to `type` `file`. Either passing no `type` altogether or passing `type` `dir` will result in an error.